[FX][resurrect] wrap - put arbitrary functions in the graph

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50161 [FX] Make len a wrapped function
* **#49753 [FX][resurrect] wrap - put arbitrary functions in the graph**

Resurrecting https://github.com/pytorch/pytorch/pull/46943

This adds a @Wrap decorator that can be put on a function. When called during
symbolic tracing, the function will be inserted in the graph rather than
when executing. When called outside of tracing, the function will execute
normally.

Differential Revision: [D25684565](https://our.internmc.facebook.com/intern/diff/D25684565)

Commit:
[FX][resurrect] wrap - put arbitrary functions in the graph

[ghstack-poisoned]

mrzzd commented on torch/fx/symbolic_trace.py:
```diff
@@ -307,13 +333,48 @@ def forward(*args, **kwargs):
+            # Install _wrap_fn_impl so that we can properly record calls to
+            # leaf functions, i.e. functions decorated with @fx.wrap
+            global _wrap_fn
+            _wrap_fn = _wrap_fn_impl
```
I wonder if we could make _wrap_fn_impl a class method. That would have two benefits: 
One is that find_proxy will not be needed anymore (since _wrap_fn_impl bounds to the proxy object). Although, you may still want to keep it to distinguish cases where no proxy is used .
But more importantly, this enables overriding _wrap_fn_impl in tracer subclasses. We already have a usecase that requires overriding _wrap_fn_impl (to enable support for symtrace).

Commit:
Update on "[FX][resurrect] wrap - put arbitrary functions in the graph"


Resurrecting https://github.com/pytorch/pytorch/pull/46943

This adds a @Wrap decorator that can be put on a function. When called during
symbolic tracing, the function will be inserted in the graph rather than
when executing. When called outside of tracing, the function will execute
normally.

Differential Revision: [D25684565](https://our.internmc.facebook.com/intern/diff/D25684565)

[ghstack-poisoned]

jamesr66a commented:
Superseded by https://github.com/pytorch/pytorch/pull/50182
